File: hash.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

// Convert one hex digit to int
int hexDigitToInt(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    c = toupper(c);
    return 10 + (c - 'A');
}

int main() {
    char password[100];
    printf("Enter your password: ");
    scanf("%99s", password);

    // Convert password to hex string
    char hexStr[200] = "";
    for (int i = 0; i < strlen(password); i++) {
        char buf[3];
        sprintf(buf, "%02X", (unsigned char)password[i]);
        strcat(hexStr, buf);
    }

    printf("Password in hex: %s\n", hexStr);

    // Some Ai (GPT)
    
    // Each 3 hex digits = 1 pixel
    int numPixels = (strlen(hexStr) + 2) / 3; // round up
    int width = numPixels;
    int height = 1;

    FILE *f = fopen("password.ppm", "w");
    if (!f) {
        printf("Error: could not create image\n");
        return 1;
    }

    // PPM header
    fprintf(f, "P3\n%d %d\n255\n", width, height);

    // Process hex string in chunks of 3
    for (int i = 0; i < strlen(hexStr); i += 3) {
        char h1 = hexStr[i];
        char h2 = (i + 1 < strlen(hexStr)) ? hexStr[i+1] : '0';
        char h3 = (i + 2 < strlen(hexStr)) ? hexStr[i+2] : '0';

        int r = hexDigitToInt(h1) * 16; // scale 0–15 → 0–240
        int g = hexDigitToInt(h2) * 16;
        int b = hexDigitToInt(h3) * 16;

        fprintf(f, "%d %d %d ", r, g, b);
    }

    fprintf(f, "\n");
    fclose(f);

    printf("Image saved as password.ppm\n");
    return 0;
}

File: hash.py
def hex_digit_to_int(c: str) -> int:
    """Convert a single hex digit (0-F) to int."""
    if c.isdigit():
        return int(c)
    return 10 + ord(c.upper()) - ord('A')


def main():
    password = input("Enter your password: ")

    # Convert password to hex string
    hex_str = "".join(f"{ord(ch):02X}" for ch in password)
    print("Password in hex:", hex_str)

    # some Ai (GPT)
    
    # Each 3 hex digits = 1 pixel
    num_pixels = (len(hex_str) + 2) // 3  # round up
    width = num_pixels
    height = 1

    with open("password.ppm", "w") as f:
        # PPM header
        f.write(f"P3\n{width} {height}\n255\n")

        # Process hex string in chunks of 3
        for i in range(0, len(hex_str), 3):
            h1 = hex_str[i]
            h2 = hex_str[i + 1] if i + 1 < len(hex_str) else "0"
            h3 = hex_str[i + 2] if i + 2 < len(hex_str) else "0"

            r = hex_digit_to_int(h1) * 16
            g = hex_digit_to_int(h2) * 16
            b = hex_digit_to_int(h3) * 16

            f.write(f"{r} {g} {b} ")

        f.write("\n")

    print("Image saved as password.ppm")


if __name__ == "__main__":
    main()
